### PR TITLE
WorkletEventHandler revamp

### PR DESCRIPTION
**Pull request by szydlovsky**
## Summary
Our `WorkletEventHandler` class (used in `useScrollViewOffset` and `useEvent` hooks as well in internal logic of `AnimatedComponent`) was made for handling only one component at a time. I made it so multiple of them can work at the same time. This way, anything based on `useEvent` (so for example `useAnimatedScrollHandler`) can be used for multiple components without needlessly copying the logic. This fixes https://github.com/software-mansion/react-native-reanimated/issues/5345 and https://github.com/software-mansion/react-native-reanimated/issues/5488 as well.

## Test plan
Effectively, this needs testing on `useAnimatedScrollHandler` and `useScrollViewOffset` as well as generally checking for any regression in AnimatedComponents.

Testing code for `useAnimatedScrollHandler` on multiple components:
<details><summary>Code</summary>

``` TYPESCRIPT
import * as React from 'react';
import { Text, View, Button, StyleSheet } from 'react-native';
import Animated, { useAnimatedScrollHandler } from 'react-native-reanimated';

type ItemValue = { title: string };
const items: ItemValue[] = [...new Array(101)].map((_each, index) => {
  return { title: `${index}` };
});

type ItemProps = { title: string };
const Item = ({ title }: ItemProps) => (
  <View style={styles.item}>
    <Text style={styles.title}>{title}</Text>
  </View>
);

export default function EmptyExample() {
  const [attachBoth, setAttachBoth] = React.useState(false);
  const [onlyOnScroll, setOnlyOnScroll] = React.useState(true);

  // props to pass into children render functions
  const onScroll = useAnimatedScrollHandler(
    onlyOnScroll
      ? {
          onScroll(e) {
            'worklet';
            console.log(e.eventName);
          },
        }
      : {
          onScroll(e) {
            'worklet';
            console.log(e.eventName);
          },
          onBeginDrag(e) {
            'worklet';
            console.log(e.eventName);
          },
          onEndDrag(e) {
            'worklet';
            console.log(e.eventName);
          },
          onMomentumBegin(e) {
            'worklet';
            console.log(e.eventName);
          },
          onMomentumEnd(e) {
            'worklet';
            console.log(e.eventName);
          },
        }
  );

  return (
    <View style={styles.container}>
      <Button
        title={`Swtich to attach ${attachBoth ? 'first' : 'both'}`}
        onPress={() => {
          setAttachBoth(!attachBoth);
        }}
      />
      <Button
        title={`Swtich events to ${
          onlyOnScroll ? 'all events' : 'only onScroll'
        }`}
        onPress={() => {
          setOnlyOnScroll(!onlyOnScroll);
        }}
      />
      <Animated.FlatList
        onScroll={onScroll}
        style={styles.listA}
        data={items}
        renderItem={({ item }) => <Item title={item.title} />}
        keyExtractor={(item) => `A:${item.title}`}
      />
      <Animated.FlatList
        onScroll={attachBoth ? onScroll : null}
        style={styles.listB}
        data={items}
        renderItem={({ item }) => <Item title={item.title} />}
        keyExtractor={(item) => `B:${item.title}`}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
  },
  listA: {
    flex: 1,
    backgroundColor: '#5F9EA0',
  },
  listB: {
    flex: 1,
    backgroundColor: '#7FFFD4',
  },
  item: {
    backgroundColor: '#F0FFFF',
    alignItems: 'center',
    padding: 20,
    marginVertical: 8,
    marginHorizontal: 16,
    borderRadius: 20,
  },
  title: {
    fontSize: 32,
  },
});
```
</details>

Testing code for `useScrollViewOffset` switching between components:
<details><summary>Code</summary>

``` TYPESCRIPT
import React, { useState } from 'react';
import Animated, {
  useAnimatedRef,
  useDerivedValue,
  useSharedValue,
  useScrollViewOffset,
} from 'react-native-reanimated';
import { Button, StyleSheet, Text, View } from 'react-native';

export default function ScrollViewOffsetExample() {
  const aref = useAnimatedRef<Animated.ScrollView>();
  const bref = useAnimatedRef<Animated.ScrollView>();
  const scrollHandler = useSharedValue(0);
  const [scrollAMounted, setScrollAMounted] = useState(true);
  const [scrollBMounted, setScrollBMounted] = useState(true);
  const [scrollAPassed, setScrollAPassed] = useState(true);

  useDerivedValue(() => {
    console.log(scrollHandler.value);
  });

  const onAMountPress = () => {
    setScrollAMounted(!scrollAMounted);
  };
  const onBMountPress = () => {
    setScrollBMounted(!scrollBMounted);
  };
  const onPassTogglePress = () => {
    setScrollAPassed(!scrollAPassed);
  };

  useScrollViewOffset(scrollAPassed ? aref : bref, scrollHandler);

  return (
    <>
      <View style={styles.positionView}>
        <Text>Test</Text>
      </View>
      <View style={styles.divider} />
      <Button
        title={`${scrollAMounted ? 'Dismount' : 'Mount'} scroll A`}
        onPress={onAMountPress}
      />
      <Button
        title={`${scrollBMounted ? 'Dismount' : 'Mount'} scroll B`}
        onPress={onBMountPress}
      />
      <Button
        title={`Toggle the ref, currently passed to ${
          scrollAPassed ? 'scroll A' : 'scroll B'
        }`}
        onPress={onPassTogglePress}
      />
      {scrollAMounted ? (
        <Animated.ScrollView
          ref={aref}
          style={[styles.scrollView, { backgroundColor: 'purple' }]}>
          {[...Array(100)].map((_, i) => (
            <Text key={i} style={styles.text}>
              A: {i}
            </Text>
          ))}
        </Animated.ScrollView>
      ) : null}
      {scrollBMounted ? (
        <Animated.ScrollView
          ref={bref}
          style={[styles.scrollView, { backgroundColor: 'lime' }]}>
          {[...Array(100)].map((_, i) => (
            <Text key={i} style={styles.text}>
              B: {i}
            </Text>
          ))}
        </Animated.ScrollView>
      ) : null}
    </>
  );
}

const styles = StyleSheet.create({
  positionView: {
    margin: 20,
    alignItems: 'center',
  },
  scrollView: {
    flex: 1,
    width: '100%',
  },
  text: {
    fontSize: 50,
    textAlign: 'center',
  },
  divider: {
    backgroundColor: 'black',
    height: 1,
  },
});
```
</details>

Testing code for multiple `useAnimatedScrollHandlers` and changing them on one component:
<details><summary>Code</summary>

``` TYPESCRIPT
import * as React from 'react';
import { Text, View, Button, StyleSheet } from 'react-native';
import Animated, { useAnimatedScrollHandler } from 'react-native-reanimated';

type ItemValue = { title: string };
const items: ItemValue[] = [...new Array(101)].map((_each, index) => {
  return { title: `${index}` };
});

type ItemProps = { title: string };
const Item = ({ title }: ItemProps) => (
  <View style={styles.item}>
    <Text style={styles.title}>{title}</Text>
  </View>
);

export default function ScrollViewExample() {
  const [attachScroll, setAttachScroll] = React.useState(true);
  const [attachMomentum, setAttachMomentum] = React.useState(false);
  const [changeScroll, setChangeScroll] = React.useState(false);
  const [changeMomentum, setChangeMomentum] = React.useState(false);

  const onScrollHandler = useAnimatedScrollHandler({
    onScroll(e) {
      console.log(e.eventName);
    },
  });

  const onMomentumHandler = useAnimatedScrollHandler({
    onMomentumBegin(e) {
      console.log(e.eventName);
    },
  });

  function reverseString(str: string): string {
    'worklet';
    return str.split('').reduce((reversed, char) => char + reversed, '');
  }

  const onScrollReversedHandler = useAnimatedScrollHandler({
    onScroll(e) {
      console.log(reverseString(e.eventName));
    },
  });

  const onMomentumReversedHandler = useAnimatedScrollHandler({
    onMomentumBegin(e) {
      console.log(reverseString(e.eventName));
    },
  });

  return (
    <View style={styles.container}>
      <Button
        title={`${attachScroll ? 'Detach' : 'Attach'} scroll handler`}
        onPress={() => {
          setAttachScroll(!attachScroll);
        }}
      />
      <Button
        title={`${attachMomentum ? 'Detach' : 'Attach'} momentum handler`}
        onPress={() => {
          setAttachMomentum(!attachMomentum);
        }}
      />
      <Button
        title={`${changeScroll ? 'Un-reverse' : 'Reverse'} scroll handler`}
        onPress={() => {
          setChangeScroll(!changeScroll);
        }}
      />
      <Button
        title={`${changeMomentum ? 'Un-reverse' : 'Reverse'} momentum handler`}
        onPress={() => {
          setChangeMomentum(!changeMomentum);
        }}
      />
      <Animated.FlatList
        onScroll={
          attachScroll
            ? changeScroll
              ? onScrollReversedHandler
              : onScrollHandler
            : null
        }
        onMomentumScrollBegin={
          attachMomentum
            ? changeMomentum
              ? onMomentumReversedHandler
              : onMomentumHandler
            : null
        }
        style={styles.list}
        data={items}
        renderItem={({ item }) => <Item title={item.title} />}
        keyExtractor={(item) => `A:${item.title}`}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
  },
  list: {
    flex: 1,
    backgroundColor: '#5F9EA0',
  },
  item: {
    backgroundColor: '#F0FFFF',
    alignItems: 'center',
    padding: 20,
    marginVertical: 8,
    marginHorizontal: 16,
    borderRadius: 20,
  },
  title: {
    fontSize: 32,
  },
});
```
</details>

Regarding Animated Components, @Latropos currently works on some tests to make it easier.

Showcase of the last example:

https://github.com/software-mansion/react-native-reanimated/assets/77503811/6501574e-7e3b-4474-8c57-0bdf7b3f4481

### CODE REVIEW

**Comment by tjzel** [comment]
Please don't merge until https://github.com/software-mansion/react-native-reanimated/pull/5790 is merged.

File: src/reanimated2/WorkletEventHandler.ts
Comment (tjzel) [inline src/reanimated2/WorkletEventHandler.ts:16]: Let's move type definitions to the bottom of the file, following the simple rule - the most important code on top.

File: src/createAnimatedComponent/createAnimatedComponent.tsx
Comment (tjzel) [inline src/createAnimatedComponent/createAnimatedComponent.tsx:259]: We already have a `_viewTag` member in the class. Why don't we use it here?

File: src/createAnimatedComponent/createAnimatedComponent.tsx
Comment (tjzel) [inline src/createAnimatedComponent/createAnimatedComponent.tsx:271]: Here you assume that there can be only one handler in a component. What if we had two separate handlers for `onScroll` and `onMomentumBegin`? This function needs a lot of improvement 😈

File: src/createAnimatedComponent/createAnimatedComponent.tsx
Comment (tjzel) [inline src/createAnimatedComponent/createAnimatedComponent.tsx:295]: Same comment as above about the viewTag.

File: src/reanimated2/WorkletEventHandler.ts
Comment (tjzel) [inline src/reanimated2/WorkletEventHandler.ts:57]: I wonder if we actually want to have it as a flattened array 🤔

**Review by tjzel — CHANGES_REQUESTED** [review]
First part of the review is here - we definitely need to address components with multiple handlers.

File: src/createAnimatedComponent/createAnimatedComponent.tsx
Comment (szydlovsky) [reply src/createAnimatedComponent/createAnimatedComponent.tsx:271]: @tjzel  This PR aims to allow multiple components for one handler, not the other way (which also wasn't there before). However, I might try adding it if You feel like it is important.

File: src/reanimated2/WorkletEventHandler.ts
Comment (szydlovsky) [reply src/reanimated2/WorkletEventHandler.ts:57]: @tjzel You might have a point, the only time `registrations` are read is when we want to unregister some events for a specific viewTag. I will try to make it into a map with viewTag keys and an array of corresponding registrationID's.

File: src/createAnimatedComponent/createAnimatedComponent.tsx
Comment (tjzel) [reply src/createAnimatedComponent/createAnimatedComponent.tsx:271]: I'm pretty sure you could pass two handlers to a single component before.

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:19]: We usually prefix private fields with underscore but I like the idea that we start using this syntax.

File: src/createAnimatedComponent/createAnimatedComponent.tsx
Comment (tomekzaw) [inline src/createAnimatedComponent/createAnimatedComponent.tsx:262]: Does it make sense to move this method where `_reattachNativeEvents` used to be for easier review and tracking changes in git blame?

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:39]: just fyi, `tag` here will be just an index (0, 1, ...), is it okay?

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:43]: Instead of storing an empty array, does it make sense to remove the key?

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:112]: Not a big fan of using `reduce` here, perhaps `forEach` would be more readable I think?

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:145]: Can we move this code where it originally used to be?

File: src/reanimated2/WorkletEventHandler.ts
Comment (szydlovsky) [reply src/reanimated2/WorkletEventHandler.ts:19]: Props to @tjzel for the idea 😄

File: src/reanimated2/WorkletEventHandler.ts
Comment (szydlovsky) [reply src/reanimated2/WorkletEventHandler.ts:39]: @tomekzaw oh, won't it be the map's key?

File: src/reanimated2/WorkletEventHandler.ts
Comment (szydlovsky) [reply src/reanimated2/WorkletEventHandler.ts:43]: @tomekzaw We might even not clean it whatsoever, since it gets overwritten a few lines later

File: src/reanimated2/WorkletEventHandler.ts
Comment (szydlovsky) [reply src/reanimated2/WorkletEventHandler.ts:145]: https://github.com/software-mansion/react-native-reanimated/pull/5845#discussion_r1541601438 who wins 😄

File: src/createAnimatedComponent/createAnimatedComponent.tsx
Comment (tjzel) [reply src/createAnimatedComponent/createAnimatedComponent.tsx:280]: I think this function wasn't complex enough to really care about it in git blame, but we can revert the name change if you insist.

File: src/reanimated2/WorkletEventHandler.ts
Comment (tjzel) [inline src/reanimated2/WorkletEventHandler.ts:37]: Even though it's private API we allow the access to `worklet` for Reanimated users since for now there's no other reliable way to merge our event handlers. Let's make it public and we'll make it private once we add support for handler merging.

**Review by tjzel — COMMENTED** [review]
Looks good now, once we handle the blocker I think it'll be mergeleable.

File: src/reanimated2/WorkletEventHandler.ts
Comment (szydlovsky) [reply src/reanimated2/WorkletEventHandler.ts:37]: Sure thing

File: src/createAnimatedComponent/createAnimatedComponent.tsx
Comment (szydlovsky) [reply src/createAnimatedComponent/createAnimatedComponent.tsx:313]: @tjzel checked - seems like viewTag doesn't change.

File: src/reanimated2/hook/useScrollViewOffset.ts
Comment (tjzel) [inline src/reanimated2/hook/useScrollViewOffset.ts:103]: This will crash if `animated.getTag()` will return a null. You can copy a guard from:

- #5862
 
to make it safer.

**Review by tjzel — CHANGES_REQUESTED** [review]

**Review by tjzel — APPROVED** [review]
Let's wait for @tomekzaw's review (one day)

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:52]: Just wondering if we can skip `_tag` here at all
```suggestion
    this.#registrations.forEach((registrationIDs) => {
```

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:63]: We can skip `return` here I think
```suggestion
      const newRegistrations = this.eventNames.map((eventName) => registerEventHandler(this.worklet, eventName, tag));
```

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:55]: This is purely subjective but no need to add two more lines of code
```suggestion
      registrationIDs.forEach((id) => unregisterEventHandler(id));
```

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:73]: same as above

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:10]: this newline was ok

File: src/reanimated2/WorkletEventHandler.ts
Comment (tomekzaw) [inline src/reanimated2/WorkletEventHandler.ts:35]: do we ever call this constructor without passing `eventNames`? why should we support that?

**Review by tomekzaw — APPROVED** [review]
LGTM, left a couple of suggestions but all of them are optional

File: src/reanimated2/WorkletEventHandler.ts
Comment (piaskowyk) [inline src/reanimated2/WorkletEventHandler.ts:75]: What happens if event names have been provided, but someone calls `registerForEvents` with the `fallbackEventName` parameter? 🤔

Comment (tjzel) [reply src/reanimated2/WorkletEventHandler.ts:75]: 🤔

File: src/reanimated2/hook/useScrollViewOffset.ts
Comment (gaearon) [inline src/reanimated2/hook/useScrollViewOffset.ts:112]: Note generally reading a mutable reference in effect cleanup is risky because it could've changed by the time you got to cleanup (e.g. it could've been ... cleaned up). Or replaced by a newer reference. So it's a good idea to do

```js
useEffect(() => {
  const instance = someRef.current
  doStuff(instance)
  return () => {
    undoStuff(instance)
  })
})
```

Comment (szydlovsky) [reply src/reanimated2/hook/useScrollViewOffset.ts:112]: Yes, thanks for pointing out - luckily newer versions have already made it this way 😄

**Pull request merged**
